src/main/common/time.c

```diff
@@ -0,0 +1,249 @@
+    unsigned int minute = dt->minutes;  // 0-59
+    unsigned int hour = dt->hours;      // 0-23
+    unsigned int day = dt->day - 1;     // 0-30
+    unsigned int month = dt->month - 1; // 0-11
+    unsigned int year = dt->year - UNIX_REFERENCE_YEAR; // 0-99
+    int32_t unixTime = (((year / 4 * (365 * 4 + 1) + days[year % 4][month] + day) * 24 + hour) * 60 + minute) * 60 + second;
+    return rtcTimeMake(unixTime, dt->millis);
+}
+
+static void rtcTimeToDateTime(dateTime_t *dt, rtcTime_t t)
+{
+    int32_t unixTime = t / MILLIS_PER_SECOND;
+    dt->seconds = unixTime % 60;
+    unixTime /= 60;
+    dt->minutes = unixTime % 60;
+    unixTime /= 60;
+    dt->hours = unixTime % 24;
+    unixTime /= 24;
+
+    unsigned int years = unixTime / (365 * 4 + 1) * 4;
+    unixTime %= 365 * 4 + 1;
+
+    unsigned int year;
+    for (year = 3; year > 0; year--) {
+        if (unixTime >= days[year][0]) {
+            break;
+        }
+    }
+
+    unsigned int month;
+    for (month = 11; month > 0; month--) {
+        if (unixTime >= days[year][month]) {
+            break;
+        }
+    }
+
+    dt->year = years + year + UNIX_REFERENCE_YEAR;
+    dt->month = month + 1;
+    dt->day = unixTime - days[year][month] + 1;
+    dt->millis = t % MILLIS_PER_SECOND;
+}
+
+static void rtcGetDefaultDateTime(dateTime_t *dateTime)
+{
+    dateTime->year = 0;
+    dateTime->month = 1;
+    dateTime->day = 1;
+    dateTime->hours = 0;
+    dateTime->minutes = 0;
+    dateTime->seconds = 0;
+    dateTime->millis = 0;
+}
+
+static bool rtcIsDateTimeValid(dateTime_t *dateTime)
+{
+    return (dateTime->year >= UNIX_REFERENCE_YEAR) &&
+           (dateTime->month >= 1 && dateTime->month <= 12) &&
+           (dateTime->day >= 1 && dateTime->day <= 31) &&
+           (dateTime->hours <= 23) &&
+           (dateTime->minutes <= 59) &&
+           (dateTime->seconds <= 59) &&
+           (dateTime->millis <= 999);
+}
+
+static void dateTimeWithOffset(dateTime_t *dateTimeOffset, dateTime_t *dateTimeInitial, int16_t minutes)
+{
+    rtcTime_t initialTime = dateTimeToRtcTime(dateTimeInitial);
+    rtcTime_t offsetTime = rtcTimeMake(rtcTimeGetSeconds(&initialTime) + minutes * 60, rtcTimeGetMillis(&initialTime));
+    rtcTimeToDateTime(dateTimeOffset, offsetTime);
+}
+
+static bool dateTimeFormat(char *buf, dateTime_t *dateTime, int16_t offset)
+{
+    dateTime_t local;
+
+    int tz_hours = 0;
+    int tz_minutes = 0;
+    bool retVal = true;
+
+    // Apply offset if necessary
+    if (offset != 0) {
+        tz_hours = offset / 60;
+        tz_minutes = ABS(offset % 60);
+        dateTimeWithOffset(&local, dateTime, offset);
+        dateTime = &local;
+    }
+
+    if (!rtcIsDateTimeValid(dateTime)) {
+        rtcGetDefaultDateTime(&local);
+        dateTime = &local;
+        retVal = false;
+    }
+
+    // XXX: Changes to this format might require updates in
+    // dateTimeSplitFormatted()
+    tfp_sprintf(buf, "%04u-%02u-%02uT%02u:%02u:%02u.%03u%c%02d:%02d",
```

Comment:
Presumably this is ISO_8601 format, https://en.wikipedia.org/wiki/ISO_8601 . If so, add a comment that states this.